### PR TITLE
Log subproceses stdout and stderr in temp files

### PR DESCRIPTION
**Pull request by MVrachev**
**Fixes issues #**: https://github.com/theupdateframework/tuf/issues/1039, https://github.com/theupdateframework/tuf/issues/1119

**Description of the changes being introduced by the pull request**:

**This pr replaces pr https://github.com/theupdateframework/tuf/pull/1041.**

Logging the stdout and stderr from the test subprocesses into
temporary files clean the console from unnecessary messages from
the server-side such as "code 404, message File not found" or
"GET" queries.

The unit tests are executed in sequential order and that's why
we can reuse one temporary file object for multiple tests.

I have used this pr to remove some redundant functions, comments, and variables.

For this pr I received great help from @jku!
Thanks!

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by MVrachev** [comment]
On a successful test run the console logs look like this: 
image

I wanted to simulate an error in the subprocess and I changed the name of the server script from `simple_server.py` to `tsimple_server.py` and here how the error was logged: 
image

**Comment by jku** [comment]
Two comments from a quick test:
* the child process output does get buffered for me (so `python3 test_updater.py TestUpdater.test_4_refresh` does not notice the 404s): we should start the server using  "python -u simple_server.py" to prevent buffering
* the server output should not be warning because warnings are printed by default. Also we don't really know what the messages are and mostly do not care. info or debug level maybe?

**Comment by MVrachev** [comment]
> * the child process output does get buffered for me (so `python3 test_updater.py TestUpdater.test_4_refresh` does not notice the 404s): we should start the server using  "python -u simple_server.py" to prevent buffering

Yes, I checked that and you are right.
The child process buffers.

The solution that I propose is to use **_two temporary files_**: one for stdout and one for stderr.
I imagine that there could be an option which will give you control if you want to see the stdout or not.
I propose that by the default, if the stderr temp file has any information in it, then we would print it and the stdout file will be ignored.


> * the server output should not be a warning because warnings are printed by default. Also, we don't really know what the messages are and mostly do not care. info or debug level maybe?

If we separate stderr from stdout as I proposed above, then we can use `logger.warning()` or `logger.error()` for stderr and `logger.debug()` or `logger.info()` for stdout.
My opinion is that `logger.error()` is appropriate for `stderr` and `logger.debug()` for stdout.

My bigger concern with that approach is how we are going to propagate the cmd option when using tox or even calling the tests with `python aggregate_tests.py`...

**Comment by jku** [comment]
> The solution that I propose is to use two temporary files: one for stdout and one for stderr.

You can do that but I don't think it's necessarily worth the effort: 
* what simple_server.py considers stderr/stdout might not map to our categories anyway (as an example: we might expect the server to print errors in some situations: then it's not an error for us)
* all of simple_server.py output is mostly debugging aid anyway: We'll want to look at it only when something isn't going the way we expected
* if you separate the streams, you no longer know if the stderr line came before the stdout line...

I think dumping all the server output to debug or info is fine

**Comment by jku** [comment]
You know what: we should think about combining (or re-ordering) this PR and issue #1111 .

If we had this in utils.py (edited after chat with Martin):

```
class TestServerProcess():
 
  # starts server, sets up the TempFile, waits for server response before returning
  def __init__(self, server='simple_server.py', port, extra_args=None, timeout=10)
  
  # logs contents of TempFile with given logger, truncates buffer
  def flush_log(self, logger)
  
  # handles kill() and communicate() on the subprocess
  def kill(self)
```
 I think this PR could be vastly smaller and easier to modify later, AND we would have the infra needed by #1111

**Comment by MVrachev** [comment]
I am working to create an abstraction that will handle the subprocess and the temporary files used for logging.
While I am doing this I will release a couple of commits and I will mark this pr as a draft before it's ready for full review.

**Comment by MVrachev** [comment]
I have pushed the initial set of commits in which I created the class which handles the server subprocess and the temporary files used for logging.

I have decided to use `logging.info` for logging in the `flus_log` function because stdout and stderr are logged together and thus it seems like a reasonable default value.

On a successful test right now no logs like `code 404, message File not found` or "GET" queries are reported.

I am planning to push at least 1 commit more in which I will remove or reword some of the test comments.

File: tests/test_download.py
Comment (jku) [inline tests/test_download.py]: port number typo -- if this did not come up in tests, maybe check why

File: tests/utils.py
Comment (jku) [inline tests/utils.py]: We already discussed but I'll leave my opinion here as well for transparency: I think this does things it does not need to... 
* We don't print test names in any other logging either: if it's not clear which test is running, why not improve that in general instead of just for server output
* if the log message from the server does not look good, I think we should fix the test server output instead of replacing things in the output here
* the output missing the server name however: e.g. "simple_server.py": this would be important sometimes

these are opinions though, feel free to disregard

File: tests/test_download.py
Comment (jku) [inline tests/test_download.py]: Hmmm, this now makes comment on line 261 untrue I think

**Comment by jku** [comment]
Looks good, thanks! I'll take this for a test run later to see what the output looks like

File: tests/test_download.py
Comment (MVrachev) [reply tests/test_download.py]: I am planning to add a function `_generate_port()` in the TestServerProcess class where I will add a function that generates a port and checks if the port is busy.
So, this code will be removed.

File: tests/test_download.py
Comment (MVrachev) [reply tests/test_download.py]: It is not shown in the tests because we are checking that there is an SSL error and not what exact type is:
```
with self.assertRaises(requests.exceptions.SSLError):
        download.unsafe_download(expired_https_url, target_data_length)
```

File: tests/utils.py
Comment (MVrachev) [reply tests/utils.py]: > We don't print test names in any other logging either: if it's not clear which test is running, why not improve that in general instead of just for server output

I think the `test_name` argument can be removed because as we discussed on a test failure the logs are showed once before all tests and a second time just after the test name:
image

I don't think we can add the test names to all logs because many of the logs are coming from public API functions and there is no need to add `test_name` argument just for that.

> if the log message from the server does not look good, I think we should fix the test server output instead of replacing things in the output here

Well, yes, maybe you are right. I will remove that part.

> the output missing the server name however: e.g. "simple_server.py": this would be important sometimes

Will add that.

**Comment by MVrachev** [comment]
I don't know why but when I run `python test_arbitrary_package_attack.py` I get a resource warning even though I have closed the `temporary_log_file` in the `clean` function: 
`/home/mvrachev/src/tuf/.editable_both/lib/python3.6/posixpath.py:86: ResourceWarning: unclosed file <_io.FileIO name=5 mode='rb+' closefd=True>
  for b in map(os.fspath, p):`

Same for every file where I am using this new abstraction...

**Comment by jku** [comment]
> I don't know why but when I run `python test_arbitrary_package_attack.py` I get a resource warning even though I have closed the `temporary_log_file` in the `clean` function:
> `/home/mvrachev/src/tuf/.editable_both/lib/python3.6/posixpath.py:86: ResourceWarning: unclosed file <_io.FileIO name=5 mode='rb+' closefd=True> for b in map(os.fspath, p):`

I'm guessing this is not _your_ unclosed file -- updater currently leaves unclosed files on pretty much every "error" situation I think. Maybe check if same warning appears on develop branch? If it does then you don't need to handle it here.

You can also get more details on the warning with` PYTHONTRACEMALLOC=50` if you want

**Comment by MVrachev** [comment]
> > I don't know why but when I run `python test_arbitrary_package_attack.py` I get a resource warning even though I have closed the `temporary_log_file` in the `clean` function:
> > `/home/mvrachev/src/tuf/.editable_both/lib/python3.6/posixpath.py:86: ResourceWarning: unclosed file <_io.FileIO name=5 mode='rb+' closefd=True> for b in map(os.fspath, p):`
> 
> I'm guessing this is not _your_ unclosed file -- updater currently leaves unclosed files on pretty much every "error" situation I think. Maybe check if same warning appears on develop branch? If it does then you don't need to handle it here.
> 
> You can also get more details on the warning with` PYTHONTRACEMALLOC=50` if you want


Yes, those warnings are visible in the develop branch too.

File: tests/utils.py
Comment (jku) [inline tests/utils.py]: I still have issues with this bit:
* very verbose: most of the time the log itself is a single line: this prints 4 times as much
* should not print "INFO: ": this is what the log level is for. If the user wants it printed, they will use a log formatter that prints the log level
* I think the earlier version of a single log entry for the server log was better (it can still be multiline) as that can be made to clearly show the beginning and end of server log without need of extra lines of "************"  (I'll paste the example I tested with below) -- do you see a problem with this approach?

```
    # Build a single log entry with multiline content
    title = "Test server (" + self.server + ") output:"
    message = [title] + log_message.splitlines()
    self.__logger.info('\n| '.join(message))
```

This way the normal oneliner now looks like this (assuming basicConfig default formatter)

```
INFO:logger_name:Test server (simple_server.py) output:
| code 404, message File not found
```

File: tests/test_proxy_use.py
Comment (jku) [inline tests/test_proxy_use.py]: This has the same issue as `test_https_connection()` -- there is now a (very small) chance of collision with cls.http_server_handler.port.

I think the real solution is to make simple_server actually report back that it did succeed in binding the port (so TestServerProcess can retry if it did not). It should probably (?) not be part of this PR though: you'll have to choose an interim solution here: I think I'd use `port=cls.http_server_handler.port + 1` because that is closest to current functionality

Comment (jku) [reply tests/test_proxy_use.py]: this applies to all servers below as well

File: tests/utils.py
Comment (jku) [inline tests/utils.py]: Not sure logging an extra error is useful? if the TimeoutError lacks detail (which I think it does, sorry about that) we should add an informative message to the error where its raised maybe.

**Comment by jku** [comment]
I left a few comments, otherwise I think it looks good

**Comment by trishankatdatadog** [comment]
Just wanted to say: thanks for doing all this tedious but necessary work! 🥇

File: tests/test_proxy_use.py
Comment (MVrachev) [reply tests/test_proxy_use.py]: Yes, I won't fix this in this pr, but I will return it to the previous state in both `test_proxy_use.py` and `test_download.py`.

File: tests/utils.py
Comment (MVrachev) [reply tests/utils.py]: I fixed it in an upcoming commit and removed the commit where I introduce the `INFO` prefix before each line.

File: tests/utils.py
Comment (MVrachev) [reply tests/utils.py]: I created a new commit where I remove that line and make `TimeoutError` message more comprehensive.

**Comment by MVrachev** [comment]
Together with the great help of @jku, we have been elaborating over those changes in the last couple of weeks and now I think this pr is ready for a review!

I have rebased and cleared the commit history to make more sense for a newcomer.

This pr hides the logs from the server because it's using `logger.info()` to log information from the stdout and stderr of the server subprocesses. That's why it's better to merge this pr after https://github.com/theupdateframework/tuf/pull/1145 which will provide a way to receive logs with `info` logging level. This shouldn't be a problem to receive a review.

File: tests/test_download.py
Comment (joshuagl) [inline tests/test_download.py:78]: Nice. 

I'd guess we don't even need to store the url during setup in many tests now that we can get the port number from the `server_process_handler`? 👍

File: tests/test_proxy_use.py
Comment (joshuagl) [inline tests/test_proxy_use.py]: Why not just let `TestServerProcess` generate this port number?

File: tests/test_updater.py
Comment (joshuagl) [inline tests/test_updater.py]: Nit: I'm curious why this empty line was removed?

File: tests/utils.py
Comment (joshuagl) [inline tests/utils.py]: Nit: here, and throughout `TestServerProcess`, I think this kind of comment might be more idiomatic as a docstring.
```suggestion
  def flush_log(self):
    """Logs contents from TempFile, truncates buffer"""
```

**Review by joshuagl — APPROVED** [review]
Thanks for this great set of cleanups @MVrachev. I added a few minor comments and questions, but this is pretty much ready to merge. Excellent PR.

File: tests/test_download.py
Comment (MVrachev) [reply tests/test_download.py:78]: Well, 3 out of 4 tests in `test_download.py` are using the `self.url.` 
That's why it make sense to keep it as a variable.

File: tests/test_proxy_use.py
Comment (MVrachev) [reply tests/test_proxy_use.py]: It could, I just tried to make the lines shorter and just doesn't change it.
We are planning to leave all instances of port generation outside `tests/utils.py` and it's in the best interest of us if we do that as well.
I will change it.

File: tests/test_updater.py
Comment (MVrachev) [reply tests/test_updater.py]: Honestly, it seems it's a mistake I did while I was resolving prs.
I will fix it right away.

**Comment by MVrachev** [comment]
I rebased my changes upon the latest changes in develop and changed the second commit `Log subproceses stdout and stderr in temp files` with the comments from Joshua.

**Comment by MVrachev** [comment]
I have decided that because I am working in the same code area I will use this pr to fix the issue https://github.com/theupdateframework/tuf/issues/1119 too and that's what the last commit is all about.

**Review by jku — APPROVED** [review]
LGTM. 

Maybe you are aware but I'll mention for future commits: instead of 

> For reference read:
> #1119

you can do `Fixes #1119` (or use a few other keywords) to auto-close on merge -- obviously it's useful only if you know the issue should be closed when commit is merged.

Also, the format `Fixes issue #N` does not seem to work -- this is why this PR cover letter is probably not going to trigger auto-closing either.

**Comment by MVrachev** [comment]
> LGTM.
> 
> Maybe you are aware but I'll mention for future commits: instead of
> 
> > For reference read:
> > #1119
> 
> you can do `Fixes #1119` (or use a few other keywords) to auto-close on merge -- obviously it's useful only if you know the issue should be closed when commit is merged.
> 
> Also, the format `Fixes issue #N` does not seem to work -- this is why this PR cover letter is probably not going to trigger auto-closing either.

Thanks! 
That's what motivated me to create pr https://github.com/theupdateframework/tuf/pull/1152.

File: tests/utils.py
Comment (joshuagl) [inline tests/utils.py]:
```suggestion
        Path to the server to run in the subprocess.
```
Runned is not a word, perhaps rephrase like this?

**Review by joshuagl — APPROVED** [review]
This is an excellent set of changes, thank you. I'd like to see a minor fixup of a docstring. After that, I think this is good to merge.

**Comment by joshuagl** [comment]
Looks good, thanks Martin. Could you rebase and squash/fixup b9ef099 into e7a8a78?

**Comment by MVrachev** [comment]
Updated.

**Comment by joshuagl** [comment]
Awesome work, thanks @MVrachev

**Pull request merged**
